#16427. Make MSBuildDeps generation with deployer relocatable

stgatilov opened the pull request:
Changelog: Feature: Make ``MSBuildDeps`` generation with deployer relocatable.
Docs: Omit

Close https://github.com/conan-io/conan/issues/16427

- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop2/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one.

I have copied the relativize_path logic from CMakeDeps, I can't say I fully understand its implementation though.

As far as I understand, paths should be relativized when deployer is present but kept absolute if package is in conan cache. I have not found in the code where this decision happens, but I have tested locally that paths are relative with deployer but absolute without it (even though everything is on the same disk letter).

Please advise where to add test if it is necessary. As far as I see, there are two tests which check that output with deployer is relocatable (`tests/functional/command/test_install_deploy.py`). If I should make a similar test for MSBuildDeps, where should I put it?

stgatilov committed:
#16427. Make ConanlibXyzRoot variable relative to the generated props file when deployer is used.

Copied the logic from conan\tools\cmake\cmakedeps\templates\target_data.py

stgatilov committed:
Imports in lexicographical order...

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/conan-io/conan?pullRequest=16441) <br/>All committers have signed the CLA.

memsharded reviewed (commented):
Thanks for your contribution @stgatilov 

This looks good. It would be necessary to add some test to cover this case. I can help adding those tests if you want or feel it is too much, but if you want to try, you can start from https://github.com/conan-io/conan/blob/develop2/test/functional/command/test_install_deploy.py,  maybe starting from some test like ``test_deploy_reference``

Note these tests only verify the generated file correctness, but not necessarily use the files in a real build. If you test it also for your full use case, that would be useful, but I don't think it is necessary to add a full building test to the suite to cover this.

stgatilov committed:
#16427. Added test that MSBuildDeps generates relative paths in props with deployer, and absolute paths without deployer.

Note: this test does not invoke msbuild, and thus does not depend on MSVC / Windows.

stgatilov commented:
I added a simple test.
Full build test would be a bit more reliable, but I guess can't have full-blown tests on every combination of settings/features.

Since my goal was a test without msbuild, the tests in `test\integration\toolchains\microsoft\test_msbuilddeps.py` were much more useful to me as reference.

memsharded commented on test/integration/toolchains/microsoft/test_msbuilddeps.py:
```diff
@@ -122,3 +123,52 @@ def package_info(self):
         assert "conan_liba" not in libb
+
+
+def test_msbuilddeps_relocatable():
```
```suggestion
@pytest.mark.skipif(platform.system() != "Windows", reason="Requires Windows")
def test_msbuilddeps_relocatable():
```

stgatilov replied:
I wrote a bit different comment, since the test does not require Windows in theory.
I guess it just does not work properly with POSIX paths and breaks their leading slash. It does not matter of course.

memsharded reviewed (approved):
The test is looking good enough, it would benefit from some refactors as parameterizing the test instead of a ``for``, but don't worry about it, we ca can do it later.

The only problem is the test is failing in other platforms than Windows:
```
                    # path should be absolute, since conan cache does not move with project

>                   assert os.path.isabs(value)

E                   AssertionError: assert False

E                    +  where False = <function isabs at 0x7f9e62578a40>('tmp/tmp9yzuhr1lconans/path with spaces/.conan2/p/b/libhb1f8faf955e10/p')

E                    +    where <function isabs at 0x7f9e62578a40> = <module 'posixpath' (frozen)>.isabs

E                    +      where <module 'posixpath' (frozen)> = os.path
```
It would be perfectly fine to add a ``@pytest.mark.skipif(platform.system() != "Windows", reason="Requires Windows")`` to the test so it only runs in WIndows

stgatilov committed:
Disabled test on Windows, converted to parametrized test.

memsharded commented:
Thanks very much for contributing this!
